### PR TITLE
MINIFICPP-902: Attempt improved caching with travis

### PR DESCRIPTION
**Pull request by phrocker**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -110,6 +110,14 @@ if(CCACHE_FOUND)
     message("-- Found ccache: ${CCACHE_FOUND}")
 endif(CCACHE_FOUND)
 
+if (UNIX AND NOT APPLE)
+  execute_process(COMMAND ${CMAKE_C_COMPILER} -fuse-ld=gold -Wl,--version ERROR_QUIET OUTPUT_VARIABLE ld_version)
+  if ("${ld_version}" MATCHES "GNU gold")
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fuse-ld=gold -Wl,--disable-new-dtags")
```
Comment (bakaid) [inline CMakeLists.txt]: I assume we used bfd so far. Switching linkers could potentially cause unexpected behaviours and serious errors. Should we make gold the default? Are we sure this works properly on all supported platforms?

File: .travis.yml [hunk 2]
```diff
@@ -42,6 +43,7 @@ matrix:
           - uuid-dev
           - doxygen
           - openssl
+          - ninja-build
```
Comment (bakaid) [inline .travis.yml]: I really like ninja and I am happy that it is added, but am I correct that we don't actually use it on Travis? I don't see the  -g/--useninja flag specified anywhere.

File: .travis.yml [hunk 3]
```diff
@@ -175,6 +184,5 @@ matrix:
         - package='libpcap'; [[ $(brew ls --versions ${package}) ]] && { brew outdated ${package} || brew upgrade ${package}; } || brew install ${package}
         - package='xz'; [[ $(brew ls --versions ${package}) ]] && { brew outdated ${package} || brew upgrade ${package}; } || brew install ${package}
         - package='libssh2'; [[ $(brew ls --versions ${package}) ]] && { brew outdated ${package} || brew upgrade ${package}; } || brew install ${package}
-
-script:
-  - ./bootstrap.sh -e -t && cd build  && make -j2 VERBOSE=1 && sudo make test ARGS="-j2 --output-on-failure" && make linter && make apache-rat
```
Comment (bakaid) [inline .travis.yml]: I don't see apache-rat anywhere. I am not sure if we have ever really used it, just want to make sure that we won't miss it (I won't).

File: .travis.yml [hunk 4]
```diff
@@ -54,6 +56,8 @@ matrix:
         # Establish updated toolchain as default
         - sudo unlink /usr/bin/gcc && sudo ln -s /usr/bin/gcc-4.8 /usr/bin/gcc
         - sudo unlink /usr/bin/g++ && sudo ln -s /usr/bin/g++-4.8 /usr/bin/g++
+      script:
+        - ./bootstrap.sh -e -t && cd build  && cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_RULE_MESSAGES=OFF .. && cmake --build . --parallel 3 && sudo make test ARGS="-j4 --output-on-failure" && make linter
```
Comment (bakaid) [inline .travis.yml]: Why is it --parallel 3 here? It is 4 everywhere else.

File: .travis.yml [hunk 5]
```diff
@@ -82,55 +87,57 @@ matrix:
         # Establish updated toolchain as default
         - sudo unlink /usr/bin/gcc && sudo ln -s /usr/bin/gcc-4.8 /usr/bin/gcc
         - sudo unlink /usr/bin/g++ && sudo ln -s /usr/bin/g++-4.8 /usr/bin/g++
+      script:
+        - ./bootstrap.sh -e -t && cd build  && cmake -DUSE_SHARED_LIBS= .. && make -j4 VERBOSE=1  && sudo make test ARGS="-j2 --output-on-failure"
```
Comment (bakaid) [inline .travis.yml]: Why can't we use cmake --build instead of make here (and in the next one) too?

File: .travis.yml (hunk 3 above)
Comment (phrocker) [reply .travis.yml]: We do not and it caused many more issues than it solved. We should validate that before release.

File: .travis.yml (hunk 2 above)
Comment (phrocker) [reply .travis.yml]: Correct.

File: .travis.yml (hunk 5 above)
Comment (phrocker) [reply .travis.yml]: We can. i think this and your above comment are sloppy cases where I rebased across machines and didn't adequately resolve the merges. Thanks

File: CMakeLists.txt (hunk 1 above)
Comment (phrocker) [reply CMakeLists.txt]: That's actually a very good point, this was a branch with a ton of commits where I attempted various things. I'd like to keep this with a conditional.

File: .travis.yml (hunk 2 above)
Comment (phrocker) [reply .travis.yml]: We used it on Windows which helped, but there is a dependency issue I did not aim to solve here. Locally I addressed that issue but didn't see a significant speed up over cmake build ( which more apply generates external project builds in parallel as Ninja does ) -- so I will remove Ninja from the travis here. I tried many incremental changes so some were left in here like this. I will remove this line as you are correct it does not serve a purpose.

File: CMakeLists.txt (hunk 1 above)
Comment (bakaid) [reply CMakeLists.txt]: I am completely for keeping it with a conditional.

File: .travis.yml (hunk 4 above)
Comment (phrocker) [reply .travis.yml]: Same as below, just sloppy rebase/merge across machines and I haven't looked through the PR yet as my focus was travis doing better. I will fix this and everything else in hopes of getting a full green in less time. Thanks for helping!!

Comment (bakaid) [reply .travis.yml]: Thank you for doing all of this, it was really needed!

File: .travis.yml (hunk 2 above)
Comment (bakaid) [reply .travis.yml]: That's perfect.

File: .travis.yml (hunk 5 above)
Comment (bakaid) [reply .travis.yml]: To answer my own question based on the Travis results: because Trusty has cmake 2.8 and that does not support it. :(

Comment (phrocker) [reply .travis.yml]: Yeah this week I keep getting time sliced to other tickets/bugs and software projects so I had completely forgotten why and assumed it was the fault of the merge. Little did I know I just can't remember anything!

**Comment by phrocker** [comment]
+1 from community and commiters. will merge.

**Pull request closed without merging**